### DISCUSSION TITLE
Re-amalgamating and interpreting Janet (tcc-mob)

### DISCUSSION THREAD

**Discussion by CosmicToast** [janet-lang/janet#1127] [Show and tell]
I was wondering if Janet was sufficiently simple to be interpreted...
So I built a re-amalgamator to distribute janet shell as a single source file to be interpreted by tcc.
Instructions: download `janet.h`, `janet.c`, and `shell.c`.
Place this script (e.g as build.sh) in the same directory and run it.
You get out an executable script named `janet` that you can place anywhere and run - it will be interpreted and ran on the fly.
Caveat: no dynamic modules, though regular modules work just fine. Also you're limited to the "default" `/usr/local/lib` library location like this, unless you change the defines manually.
The script:
```sh
#!/bin/sh -e

# extract janet.h include location from janet.c
# we skip the include line itself using %%, which matches any line
csplit -f janet.c. -n1 -s janet.c '/janet.h/' '%%'
# we need this header to be able to execute it directly
# dynamic modules will not work with this so let's disable them
echo '#!/usr/bin/env -S tcc -DJANET_NO_DYNAMIC_MODULES -w -run' > head
# strip out any other mentions of janet.h and assemble the script
cat head janet.c.0 janet.h janet.c.1 shell.c |
	grep -v 'janet\.h' > janet
chmod +x janet

# clean up
rm -f janet.c.* head
```

It actually runs quite fast and can run some "real" scripts (again, as long as there are no native modules).
The startup time is essentially instant, my native optimized build runs `janet -e (quit)` in 9.12ms, while the tcc-interpreted one does the same in 95.39ms.
I find it to be a rather good sign in general when a program written in C (let alone an entire language) can be built by something like tcc, let alone interpreted, so this was a fairly cool ~30m experiment.

**Comment by sogaiu** [comment]
Looks neat!

Locally I didn't quite get it to work due to header file issues:

```
$ ./janet
In file included from ./src/core/abstract.c:34:
./src/core/abstract.c:34: error: include file 'stdatomic.h' not found
```

```
$ find /usr/include | grep stdatomic.h
/usr/include/c++/12/stdatomic.h
```

Will try again when I'm more awake (^^;

**Reply by CosmicToast** [reply]
There should be a `stdatomic.h` that's distributed with tcc, which it should automatically find if it's been installed correctly.
The above snippets suggests it hasn't.
You can always try building it by hand (something like `./configure --prefix=/usr/local`, make sure it's actually appropriate for your system, then make and make install).
The result should do it just fine.

**Reply by sogaiu** [reply]
Thanks for the tips.

I'm using a packaged tcc on an Ubuntu machine, so possibly that's a factor.

I tried compiling from source I got from [here](https://download.savannah.gnu.org/releases/tinycc/), but the tail end of the output was:
```
make[1]: Entering directory '/home/user/src/tcc-0.9.27/lib'
../tcc -c libtcc1.c -o libtcc1.o -B..
../tcc -c alloca86_64.S -o alloca86_64.o -B..
../tcc -c alloca86_64-bt.S -o alloca86_64-bt.o -B..
../tcc -c va_list.c -o va_list.o -B..
../tcc -c bcheck.c -o bcheck.o -B..
bcheck.c:738: error: '__malloc_hook' undeclared
make[1]: *** [Makefile:64: bcheck.o] Error 1
make[1]: Leaving directory '/home/user/src/tcc-0.9.27/lib'
make: *** [Makefile:230: libtcc1.a] Error 2
```

Some searching suggests [there was some change to glibc](https://stackoverflow.com/questions/17803456/an-alternative-for-the-deprecated-malloc-hook-functionality-of-glibc) some time back.

May be it's another case of Mr. Bellard's software bitrotting...

I guess I'll do some more digging :)

**Reply by sogaiu** [reply]
Looks like getting source from https://repo.or.cz/tinycc.git works out better.

```
$ time ./janet -e "(quit)"

real	0m0.143s
user	0m0.111s
sys	0m0.032s
```

I tried it out on [some multi-threaded code](https://gist.github.com/sogaiu/4f1094400c1ecdb5f78a94ccbf426e32) and sometimes get:

```
Segmentation fault (core dumped)
```

or:

```
C runtime error at line 629 in file ./src/core/gc.c: invalid janet_sfree
```

Hmm, may be `jane` or `jean` could work as names...

**Reply by CosmicToast** [reply]
For fun I tried running it on MacOS... it was failing to handle JANET_THREAD_LOCAL, so I added the JANET_SINGLE_THREADED define, at which point I started getting the same runtime error you're seeing.
I'm going to guess that's a tcc moment.
